specs/fault-proof.md

```diff
@@ -0,0 +1,412 @@
+
+- The client initiates a request for data by writing a `bytes32` [pre-image key](#pre-image-key-types).
+- The
+
+### Pre-image key types
+
+Pre-images are identified by a `bytes32` type-prefixed key:
+
+- The first byte identifies the type of request.
+- The remaining 31 bytes identify the pre-image key.
+
+#### Type `0`: Zero key
+
+The zero prefix is illegal. This ensures all pre-image keys are non-zero,
+enabling storage lookup optimizations and avoiding easy mistakes with invalid zeroed keys in the EVM.
+
+#### Type `1`: Local key
+
+Information specific to the dispute: the remainder of the key may be an index, a string, a hash, etc.
+Only the contract(s) managing this dispute instance may serve the value for this key:
+it is localized and contex-dependent.
+
+This type of key is used for program bootstrapping, to identify the initial input arguments by index or name.
+
+#### Type `2`: Global keccak256 key
+
+This type of key uses a global pre-image store contract, and is fully context-independent and permissionless.
+I.e. every key must have a single unique value, regardless of chain history or time.
+
+Using a global store reduces duplicate pre-image registration work,
+and avoids unnecessary contract redeployments per dispute.
+
+This global store contract should be non-upgradeable.
+
+Since `keccak256` is a safe 32-byte hash input, the first bit is overwritten with a `2` to derive the key,
+while keeping the rest of the key "readable" (matching the original hash).
+
+#### Type `3`: Global generic key
+
+Reserved. This scheme allows for unlimited application-layer pre-image types without fault-proof VM redeployments.
+
+This is a generic version of a global key store: `key = 0x03 ++ keccak256(x, sender)[1:]`, where:
+
+- `key` is a `bytes32`, which can be a hash of an arbitrary-length type of cryptographically secure commitment.
+- `sender` is a `bytes32` identifying the pre-image inserter address (left-padded to 32 bytes)
+
+This global store contract should be non-upgradeable.
+
+The global contract is permissionless: users can standardize around external contracts that verify pre-images
+(i.e. a specific `sender` will always be trusted for a specific kind of pre-image).
+The external contract the pre-image into the global store for usage by all fault proof VMs without upgrades.
+
+Users may standardize around upgradeable external pre-image contracts,
+in case the implementation of the verification of the pre-image is expected to change.
+
+The store update function is `update(key bytes32, offset uint64, span uint8, value bytes32)`:
+
+- Only part of the pre-image, starting at `offset`, and up to (incl.) 32 bytes `span` can be inserted at a time.
+- Pre-images may have an undefined length (e.g. a stream), we only need to know how many bytes of `value` are usable.
+- The key and offset will be hashed together to uniquely store the `value` and `span`, for later pre-image serving.
+
```

Comment:
Do we need to implement this specific approach to breaking down data here or can we let the global contract accept any size data and programs can choose to break it up using their own key scheme as needed?

Reply:
The VM may read the pre-image up to 32 bytes at a time, but may read less if it chooses to. The pre-image store would just always serve the full available 32 bytes at that offset, and it's up to the VM to mask out the part it didn't want to read. This simplifies the pre-image store contract to just "32 bytes at any offset in the pre-image", with a special case for padding at the end if the pre-image isn't a multiple of 32 bytes.